source/common/stats/stats_impl.h

```diff
@@ -365,6 +363,10 @@ class HeapRawStatDataAllocator : public RawStatDataAllocator {
   // RawStatDataAllocator
   RawStatData* alloc(const std::string& name) override;
   void free(RawStatData& data) override;
+  ~HeapRawStatDataAllocator();
+
+private:
+  RawStatData* data_ = nullptr;
```

Comment:
I'm a little confused about this fix, so feel free to correct me if I'm misunderstanding.  This seems to couple a single allocator with a single allocated object.

I thought this allocator was a long-lived object that allocates many `RawStatData` objects for use elsewhere.  Is it your intention to change it to an allocator per `RawStatData` since you're keeping a member variable that tracks a single `RawStatData` in the allocator?  If so, wouldn't you need to ensure that objects like `IsolatedStoreImpl` create an allocator per stat that they initialize?  Without this, you're likely leaking `RawStatData` objects (which is likely where the ASAN failures are coming from).  I'm skeptical that this is the right direction to go to fix this.

Reply:
@mrice32 FYI I think all that's needed here is just proper ref-counting per the other open ticket but I'm not sure. I don't know where the original crash was seen.

Reply:
@mrice32 Thanks for looking into pr

Original error was "calling pure virtual function ", this was coming from  destructor of  ~CounterImpl() since it was calling alloc_.free(data_).
I was trying to create object of counterimpl and passing derived class reference (HeapRawStatDataAllocator) as one of the arguments

Stats::HeapRawStatDataAllocator alloc_;
Stats::RawStatData *data = alloc_.alloc(name);
Stats::CounterSharedPtr c1(new Stats::CounterImpl(*data, alloc_, std::string(name),std::move(tags)));

since in  RawStatDataAllocator free is a pure virtual function.  It seems intention was to call free of the derived class (HeapRawStatDataAllocator) in ~counterImpl destructor but by the time we reach destructor derived class reference was already lost since derived class destructor was called before.

Also for your suggestion of ref counting. Currently expectation is ref_count will be always one so ref count is not being incremented. I got this understanding based on the existing comment in free function (// This allocator does not ever have concurrent access to the raw data.). 

Please let me know your input , i will change code accordingly.

Reply:
Great, thanks for the explanation!

Why would the `HeapRawStatDataAllocator` class's methods be gone during the call to the destructor of `CounterImpl`?  `CounterImpl just has a reference to `HeapRawStatDataAllocator`, but it doesn't appear to have any inheritance relationship with or ownership over `HeapRawStatDataAllocator`.  IIUC, the reference to `HeapRawStatDataAllocator` should be valid in the `CounterImpl` destructor as long as `HeapRawStatDataAllocator` (wherever it's owned) has not been destructed already.

One thing that might help here (credit @dnoe) would be if you could write up a test case that causes the failure because it's difficult to understand exactly what we're trying to fix if we don't have visibility into where it occurs.  If a test in #2757 triggers this, then you can just point me to that.

Reply:
@mrice32 yes test in #2757 will recreate the issue, just cherry-pick the commit, since you need change in other file as well since that UT is testing that function :).

Reply:
@pitiwari I will comment there.  I think I may see the issue.

Reply:
ok :)

Reply:
Given the discussion and work in #2757, should this PR be closed? Or is there still an issue that needs to be fixed here?

Reply:
I think @pitiwari was going to look into fixing #2453 here.  However, maybe it makes sense to close this and open a new PR when the fix is ready since it may be a week or two. @pitiwari, what do you think?

Reply:
My vote is to close this PR and create a new one to fix #2453. This one has history/comments for fixing 2 different issues, neither of which is the one that will now be fixed here.